src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -15,15 +15,41 @@
 
 namespace Microsoft.CodeAnalysis.LanguageServer.Handler
 {
-    [Export(typeof(RequestExecutionQueue)), Shared]
+    /// <summary>
+    /// Coordinates the exectution of LSP messages to ensure correct results are sent back.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// When a request comes in for some data the handler must be able to access a soluition state that is correct
+    /// at the time of the request, that takes into account any text change requests that have come in  previously
+    /// (via textDocument/didChange for example).
+    /// </para>
+    /// <para>
+    /// This class acheives this by distinguishing between mutating and non-mutating requests, and ensuring that
+    /// when a mutating request comes in, its processing blocks all subsequent requests. As each request comes in
+    /// it is added to a queue, and a queue item will not be retreived while a mutating request is running. Before
+    /// any request is handled the solution state is created by merging workspace solution state, which could have
+    /// changes from non-LSP means (eg, adding a project reference), with the current "mutated" state.
+    /// When a non-mutating work item is retrieved from the queue, it is given the current solution state, but then
+    /// run in a fire-and-forget fashion.
+    /// </para>
+    /// <para>
+    /// Regardless of whether a request is mutating or not, or blocking or not, is an implementation detail of this class
+    /// and any consumers observing the results of the task returned from <see cref="ExecuteAsync{TRequestType, TResponseType}(bool, IRequestHandler{TRequestType, TResponseType}, TRequestType, ClientCapabilities, string?, CancellationToken)"/>
+    /// will see the results of the handling of the request, whenever it occurred.
+    /// </para>
+    /// <para>
+    /// Exceptions in the handling of non-mutating requests are sent back to callers. Exceptions in the processing of
+    /// the queue will close the LSP connection so that the client can reconnect. Exceptions in the handling of mutating
+    /// requests will also close the LSP connection, as at that point the mutated solution is in an unknown state.
+    /// </para>
```

Comment:
This part of the doc comment is speculative, not actually implemented by code in this PR.

Reply:
I would prefer leaving out the docs until we actually write it